src/util/Options.js

```diff
@@ -0,0 +1,169 @@
+ * @property {WebsocketOptions} [ws] Options for the WebSocket
+ * @property {HTTPOptions} [http] HTTP options
+ */
+
+/**
+ * WebSocket options (these are left as snake_case to match the API)
+ * @typedef {Object} WebsocketOptions
+ * @property {number} [large_threshold=50] Number of members in a guild after which offline users will no longer be
+ * sent in the initial guild member list, must be between 50 and 250
+ */
+
+/**
+ * HTTP options
+ * @typedef {Object} HTTPOptions
+ * @property {number} [version=9] API version to use
+ * @property {string} [api='https://discord.com/api'] Base url of the API
+ * @property {string} [cdn='https://cdn.discordapp.com'] Base url of the CDN
+ * @property {string} [invite='https://discord.gg'] Base url of invites
+ * @property {string} [template='https://discord.new'] Base url of templates
+ * @property {Object} [headers] Additional headers to send for all API requests
+ */
+
+/**
+ * Contains various utilities for client options.
+ */
+class Options extends null {
+  /**
+   * The default client options.
+   * @returns {ClientOptions}
+   */
+  static createDefault() {
+    return {
+      shardCount: 1,
+      makeCache: this.cacheWithLimits({ MessageManager: 200 }),
+      messageCacheLifetime: 0,
+      messageSweepInterval: 0,
+      invalidRequestWarningInterval: 0,
+      partials: [],
+      restWsBridgeTimeout: 5000,
+      restRequestTimeout: 15000,
+      restGlobalRateLimit: 0,
+      retryLimit: 1,
+      restTimeOffset: 500,
+      restSweepInterval: 60,
+      presence: {},
+      ws: {
+        large_threshold: 50,
+        compress: false,
+        properties: {
+          $os: process.platform,
+          $browser: 'discord.js',
+          $device: 'discord.js',
+        },
+        version: 9,
+      },
+      http: {
+        version: 9,
+        api: 'https://discord.com/api',
+        cdn: 'https://cdn.discordapp.com',
+        invite: 'https://discord.gg',
+        template: 'https://discord.new',
+      },
+    };
+  }
+
+  /**
+   * Create a cache factory using predefined limits.
+   * @param {Partial<Record<string, number>>} [limits={}] Limits for structures.
+   * @returns {CacheFactory}
+   */
+  static cacheWithLimits(limits = {}) {
+    const Collection = require('./Collection');
+    const LimitedCollection = require('./LimitedCollection');
+
+    return manager => {
+      const limit = limits[manager.name];
+      if (limit === null || limit === undefined || limit === Infinity) {
+        return new Collection();
+      }
+      return new LimitedCollection(limit);
+    };
+  }
+
+  /**
+   * Create a cache factory that always caches everything.
+   * @returns {CacheFactory}
+   */
+  static cacheEverything() {
+    const Collection = require('./Collection');
+    return () => new Collection();
+  }
+
+  /**
+   * Create a cache factory that always caches nothing.
+   * @returns {CacheFactory}
+   */
+  static cacheNothing() {
```

Comment:
I'm not too sure how to feel about this.
This is going to "break" pretty much everything except dms.

Maybe at least add a warning about that?

Reply:
If it breaks too much for reasonable use we can just remove it, let people figure it out.